clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -517,6 +519,50 @@ public static int murmur2(final byte[] data) {
         return h;
     }
 
+    /**
+     * Generates 32 bit murmur2 hash from ByteBuffer
+     * @param data ByteBuffer to hash
+     * @return 32 bit hash of the given ByteBuffer
+     */
+    @SuppressWarnings("fallthrough")
+    public static int murmur2(ByteBuffer data) {
+        final int length = data.remaining();
+        final int seed = 0x9747b28c;
+        // 'm' and 'r' are mixing constants generated offline.
+        // They're not really 'magic', they just happen to work well.
+        final int m = 0x5bd1e995;
+        final int r = 24;
+
+        // Initialize the hash to a random value
+        int h = seed ^ length;
+        final int length4 = length / 4;
+        data = data.order() == LITTLE_ENDIAN ? data : data.slice().order(LITTLE_ENDIAN);
```

Comment:
I don't think it's necessary to do that. From this [commit](https://github.com/apache/kafka/pull/12685/commits/721db38b0580250a122f7dd178172ddec616cbe6), it seems that this change does not have any compatibility issues.

Reply:
> I don't think it's necessary to do that

can you help me understand how this works? I am concerned about the fact that murmur2 requires little endian order but incoming data could be in big endian format.

Reply:
> > I don't think it's necessary to do that
> 
> can you help me understand how this works? I am concerned about the fact that murmur2 requires little endian order but incoming data could be in big endian format.

We can approach the problem in the following way: as long as the value of variable `k` obtained in both `Utils#murmur2(ByteBuffer)` and `Utils#murmur2(byte[])` is consistent, we can ensure that the results calculated by `Utils#murmur2(ByteBuffer)` and `Utils#murmur2(byte[])` are the same. 

Next, I will use a simple example to explain why it doesn't affect the correctness of `Utils#murmur2(ByteBuffer)`whether the `ByteBuffer` passed by the user is `BIG_ENDIAN` or `LITTLE_ENDIAN` order.

```
final ByteBuffer heapBuffer = ByteBuffer.allocate(4).order(BIG_ENDIAN);
heapBuffer.putInt(1234567890); // [73, -106, 2, -46]
heapBuffer.flip();

// [73, -106, 2, -46]
final byte[] bytes = Utils.toArray(heapBuffer); 

// obtain variable k from bytes just like Utils#murmur2(byte[])
final int k0 = (bytes[0] & 0xff) + ((bytes[1] & 0xff) << 8) + ((bytes[2] & 0xff) << 16) + ((bytes[3] & 0xff) << 24);

// obtain variable k from bytes just like Utils#murmur2(ByteBuffer)
final int k1 = heapBuffer.slice().order(LITTLE_ENDIAN).getInt(0);

// It will print `k0=-771582391, k1=-771582391`
System.out.println("k0=" + k0 + ", k1=" + k1);
```

NOTE: `Utils#murmur2(ByteBuffer)` does not require the `ByteBuffer` passed by the user to be in `LITTLE_ENDIAN` order. `Utils#murmur2(ByteBuffer)` simply reads the data from the `ByteBuffer` in `LITTLE_ENDIAN` order.